### PR TITLE
ud async timer, retransmisions and pending queue

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1040/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/470/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
can you please squash ud-related commits, and along with the GTEST commit (i guess the test would fail on UD without it)? 6a8924e+58fa484+9808552
also @alinask please note that this means the counter would have to be removed from shm iface on PR #424

**Comment by shamisp** [comment]
autobot build

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1045/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/475/ for details (Mellanox internal link).

File: src/ucs/datastruct/arbiter.c [hunk 1]
```diff
@@ -146,6 +151,10 @@ void ucs_arbiter_dispatch_nonempty(ucs_arbiter_t *arbiter, unsigned per_group,
                if (elem == last_elem) {
                     /* Only element */
                     group->tail = NULL; /* Group is empty now */
+                    arbiter->group_count--;
+                    if (next_count > arbiter->group_count) {
```
Comment (yosefe) [inline src/ucs/datastruct/arbiter.c]: maybe use ucs_min?

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 2]
```diff
@@ -405,9 +429,11 @@ uct_ud_mlx5_ep_create_connected(uct_iface_h iface_h,
     ucs_trace_data("TX: CREQ (qp=%x lid=%d)", if_addr->qp_num, if_addr->lid);
     uct_ud_mlx5_ep_tx_skb(iface, ep, skb);
     uct_ud_iface_complete_tx_skb(&iface->super, &ep->super, skb);
+    uct_ud_leave(&iface->super);
     return UCS_OK;
 
 err:
+    uct_ud_leave(&iface->super);
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: looking at order of enter/create, do we need to reverse order of leave/destroy here?

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 3]
```diff
@@ -405,9 +429,11 @@ uct_ud_mlx5_ep_create_connected(uct_iface_h iface_h,
     ucs_trace_data("TX: CREQ (qp=%x lid=%d)", if_addr->qp_num, if_addr->lid);
     uct_ud_mlx5_ep_tx_skb(iface, ep, skb);
     uct_ud_iface_complete_tx_skb(&iface->super, &ep->super, skb);
+    uct_ud_leave(&iface->super);
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: maybe something like "goto out_leave"?

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 4]
```diff
@@ -481,6 +507,9 @@ static UCS_CLASS_INIT_FUNC(uct_ud_mlx5_iface_t,
                               pd, worker,
                               dev_name, rx_headroom, 0, config);
 
+    self->super.ops.progress = uct_ud_mlx5_iface_async_progress;
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: should it be called "ops.async_progress"?

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 5]
```diff
@@ -481,6 +507,9 @@ static UCS_CLASS_INIT_FUNC(uct_ud_mlx5_iface_t,
...
+    self->super.ops.tx_skb   = uct_ud_mlx5_ep_tx_ctl_skb;
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: should it be called "ops.tx_ctrl_skb"?

File: src/uct/ib/ud/base/ud_def.h [hunk 6]
```diff
@@ -167,9 +168,14 @@ enum {
     UCT_UD_EP_OP_INPROGRESS = UCS_BIT(7)   /* bit is set when op is sceduled */
 };
 
+#define UCT_UD_EP_OP_CTL (UCT_UD_EP_OP_ACK|UCT_UD_EP_OP_ACK_REQ|UCT_UD_EP_OP_RESEND|UCT_UD_EP_OP_CREP)
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_def.h]: indentation

File: src/uct/ib/ud/base/ud_ep.c [hunk 7]
```diff
@@ -18,14 +18,58 @@ static void uct_ud_ep_reset(uct_ud_ep_t *ep)
...
+    UCT_UD_EP_HOOK_CALL_TIMER(ep);
+    now = ucs_twheel_get_time(&iface->async.slow_timer);
+
+    if (ucs_queue_is_empty(&ep->tx.window)) {
+        return;
+    }
+
+    if (now - ep->tx.send_time > 3*uct_ud_slow_tick()) {
+        ucs_debug("sceduling resend now: %llu send_time: %llu diff: %llu tick: %llu",
+                  now, ep->tx.send_time, now - ep->tx.send_time, uct_ud_slow_tick());
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: ucs_trace

File: src/uct/ib/ud/base/ud_ep.c [hunk 8]
```diff
@@ -18,14 +18,58 @@ static void uct_ud_ep_reset(uct_ud_ep_t *ep)
...
+    }
+
+    ucs_wtimer_add(&iface->async.slow_timer, &ep->slow_timer, 
+                   uct_ud_slow_tick());
+}
+
+static void uct_ud_ep_fast_timer(ucs_callback_t *self)
+{
+    uct_ud_ep_t *ep = ucs_container_of(self, uct_ud_ep_t, fast_timer);
+    ucs_error("fast tick %p", ep);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: unimplemented?

File: src/uct/ib/ud/base/ud_ep.c [hunk 9]
```diff
@@ -18,14 +18,58 @@ static void uct_ud_ep_reset(uct_ud_ep_t *ep)
...
+}
+
+static void uct_ud_ep_pending_op_init(uct_ud_ep_pending_op_t *pending)
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: do we need 2 separate functions for these one-liners?

File: src/uct/ib/ud/base/ud_ep.c [hunk 10]
```diff
@@ -416,8 +445,263 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+       (!uct_ud_ep_is_connected(ep) ||
+         uct_ud_ep_no_window(ep))) {
+        return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
+    }
+
+    if (&ep->tx.pending.elem == elem) { 
+        uct_ud_ep_do_pending_ctl(ep, iface);
+        if (!uct_ud_ep_ctl_op_isany(ep)) {
+            /* no more ctl - dummy elem can be removed */
+            result = UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: i think you can avoid the variable "result" and jest return the required status directly.

File: src/uct/ib/ud/base/ud_ep.c [hunk 11]
```diff
@@ -416,8 +445,263 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+        else {
+            /* try to send ctl messages */
+            uct_ud_ep_do_pending_ctl(ep, iface);
+            result = UCS_ARBITER_CB_RESULT_NEXT_GROUP;
+        }
+    }
+    return result;
+}
+
+ucs_status_t uct_ud_ep_pending_add(uct_ep_h ep_h, uct_pending_req_t *r)
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: pls rename 'r' to 'req'

File: src/uct/ib/ud/base/ud_inl.h [hunk 12]
```diff
@@ -1,36 +1,78 @@
...
+                                    &ep->tx.pending.elem);
         ep->tx.pending.ops |= UCT_UD_EP_OP_INPROGRESS;
     }
     ep->tx.pending.ops |= op;
+    ucs_arbiter_group_schedule(&iface->tx.pending_q, &ep->tx.pending.group);
+}
+
+
+static UCS_F_ALWAYS_INLINE void
+uct_ud_ep_ctl_op_del(uct_ud_ep_t *ep, int ops)
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_inl.h]: pls add short doc to these inline functions and if possible use more clear names

File: src/uct/ib/ud/base/ud_ep.c [hunk 13]
```diff
@@ -416,8 +445,263 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+{
+    ucs_arbiter_elem_t *elem;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: this variable is not needed

File: src/uct/ib/ud/base/ud_ep.h [hunk 14]
```diff
@@ -46,10 +48,12 @@ typedef ucs_status_t (*uct_ud_ep_hook_t)(uct_ud_ep_t *ep, uct_ud_neth_t *neth);
 
 #define UCT_UD_EP_HOOK_CALL_RX(ep, neth) \
     if ((ep)->rx.rx_hook(ep, neth) != UCS_OK) { \
+        ucs_trace_data("RX FILTER: PACKET DROPPED"); \
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: maybe dump the net header here?

File: src/uct/ib/ud/base/ud_ep.h [hunk 15]
```diff
@@ -46,10 +48,12 @@ typedef ucs_status_t (*uct_ud_ep_hook_t)(uct_ud_ep_t *ep, uct_ud_neth_t *neth);
...
         return; \
     }
 
 #define UCT_UD_EP_HOOK_CALL_TX(ep, neth) (ep)->tx.tx_hook(ep, neth);
+#define UCT_UD_EP_HOOK_CALL_TIMER(ep) (ep)->timer_hook(ep, NULL);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: indentation

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1046/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/476/ for details (Mellanox internal link).

File: src/uct/ib/ud/base/ud_iface.c [hunk 16]
```diff
@@ -475,64 +504,47 @@ void uct_ud_iface_replace_ep(uct_ud_iface_t *iface,
...
-        if (status == UCS_ERR_NO_RESOURCE) {
-            return;
-        }
-        if (status == UCS_INPROGRESS) {
-            continue;
-        }
-        tx_skb(iface, ep, skb);
-        uct_ud_ep_log_tx(iface, ep, skb);
+    /* Release acknowledged skb's */
+    ucs_queue_for_each(skb, &iface->tx.res_skbs, queue) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: ucs_queue_for_each_safe

File: src/uct/ib/ud/verbs/ud_verbs.c [hunk 17]
```diff
@@ -468,8 +494,11 @@ static UCS_CLASS_INIT_FUNC(uct_ud_verbs_iface_t, uct_pd_h pd, uct_worker_h worke
     self->tx.wr_skb.sg_list           = self->tx.sge;
     self->tx.wr_skb.num_sge           = 1;
 
+    uct_ud_iface_reserve_skbs(&self->super, self->super.tx.available);
```
Comment (yosefe) [inline src/uct/ib/ud/verbs/ud_verbs.c]: can this be moved to super init (of uct_ud_iface_t)?

File: test/gtest/uct/test_ud_pending.cc [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+    req_count = 0;
+    connect();
+
+    for(i = 0; i < N; i++) {
+        r[i].func = pending_cb_busy;
+        EXPECT_EQ(UCS_OK, uct_ep_pending_add(m_e1->ep(0), &r[i]));
+    }
+    short_progress_loop();
+    /* requests must be dispatched from progress */
+    //EXPECT_EQ(0, req_count);
```
Comment (yosefe) [inline test/gtest/uct/test_ud_pending.cc]: remove this?

File: test/gtest/uct/test_ud_slow_timer.cc [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+#include "ud_base.h"
+extern "C" {
+#include <ucs/time/time.h>
+#include <ucs/datastruct/queue.h>
+#include <uct/ib/ud/base/ud_ep.h>
+#include <uct/ib/ud/base/ud_iface.h>
+};
+
+class test_ud_slow_timer : public ud_base_test {
+    public:
```
Comment (yosefe) [inline test/gtest/uct/test_ud_slow_timer.cc]: pls remove this

File: test/gtest/uct/test_ud_slow_timer.cc [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    EXPECT_EQ(0, ucs_frag_list_sn(&ep(m_e2)->rx.ooo_pkts));
+    twait(500);
+    //short_progress_loop();
+    
+    EXPECT_EQ(N+1, ep(m_e1)->tx.psn);
+    EXPECT_EQ(N, ucs_frag_list_sn(&ep(m_e2)->rx.ooo_pkts));
+}
+
+/* ack while doing retransmit */
+static int packet_count, rx_limit;
```
Comment (yosefe) [inline test/gtest/uct/test_ud_slow_timer.cc]: can we avoid these global variables?

File: test/gtest/uct/uct_test.h [hunk 21]
```diff
@@ -74,6 +75,25 @@ class uct_test : public testing::TestWithParam<const resource*>,
         void flush() const;
 
     private:
+        class async_wrapper {
+            public:
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: indentation - public/private should start on same column as 'class'

File: test/gtest/uct/uct_test.h [hunk 22]
```diff
@@ -74,6 +75,25 @@ class uct_test : public testing::TestWithParam<const resource*>,
...
+                ucs_async_context_t   m_async;
+                async_wrapper() {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: pls move function body to cpp file

File: src/uct/ib/ud/accel/ud_mlx5.c (hunk 5 above)
Comment (alex-mikheev) [inline src/uct/ib/ud/accel/ud_mlx5.c]: it is also used to resend packets

File: src/uct/ib/ud/base/ud_ep.c (hunk 8 above)
Comment (alex-mikheev) [inline src/uct/ib/ud/base/ud_ep.c]: yep. there is only slow retransmit timer in this commit

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1060/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/487/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1061/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/488/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1062/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/489/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
```
 CXX    uct/gtest-test_ud_ds.o
cc1plus: warnings being treated as errors
uct/ud_base.h: In member function 'void ud_base_test::twait(int)':
uct/ud_base.h:59: warning: converting to 'int' from 'double'
cc1plus: warnings being treated as errors
uct/ud_base.h: In member function 'void ud_base_test::twait(int)':
uct/ud_base.h:59: warning: converting to 'int' from 'double'
make[2]: *** [uct/gtest-test_ud_slow_timer.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [uct/gtest-test_ud.o] Error 1
cc1plus: warnings being treated as errors
uct/ud_base.h: In member function 'void ud_base_test::twait(int)':
uct/ud_base.h:59: warning: converting to 'int' from 'double'
make[2]: *** [uct/gtest-test_ud_pending.o] Error 1
make[2]: Leaving directory `/autofs/nccs-svm1_home1/buildbot/buildbot/pod7/ucx-rhel5-x86_64/build/test/gtest'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/autofs/nccs-svm1_home1/buildbot/buildbot/pod7/ucx-rhel5-x86_64/build'
make: *** [all] Error 2
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1064/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/491/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/492/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/493/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
looks good to me

File: src/uct/ib/ud/base/ud_ep.c [hunk 23]
```diff
@@ -416,8 +437,277 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+        /* can send more control - come here later */
+        result = UCS_ARBITER_CB_RESULT_NEXT_GROUP;
+    } else {
+        /* no more control - nothing to do in
+         * this dispatch cycle. */
+        result = UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
+    }
+    return result;
+}
+/**
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: space line

File: src/uct/ib/ud/base/ud_ep.c [hunk 24]
```diff
@@ -416,8 +437,277 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+{
+    ucs_arbiter_cb_result_t result; 
+
+    if (uct_ud_ep_ctl_op_isany(ep)) {
+        /* can send more control - come here later */
+        result = UCS_ARBITER_CB_RESULT_NEXT_GROUP;
+    } else {
+        /* no more control - nothing to do in
+         * this dispatch cycle. */
+        result = UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: can do return right here

File: src/uct/ib/ud/base/ud_ep.c [hunk 25]
```diff
@@ -416,8 +437,277 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
+       (!uct_ud_ep_is_connected(ep) ||
+         uct_ud_ep_no_window(ep))) {
+        return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
+    }
+
+    if (&ep->tx.pending.elem == elem) { 
+        uct_ud_ep_do_pending_ctl(ep, iface);
+        if (uct_ud_ep_ctl_op_isany(ep)) {
+            /* there is still some ctl left. go to next group */
+            result = UCS_ARBITER_CB_RESULT_NEXT_GROUP;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: can retrun XXX without using the status variable

File: src/uct/ib/ud/base/ud_ep.h [hunk 26]
```diff
@@ -85,7 +150,10 @@ struct uct_ud_ep {
          uct_ud_psn_t           max_psn;      /* Largest PSN that can be sent - (ack_psn + window) (from incoming packet) */
          uct_ud_psn_t           acked_psn;    /* last psn that was acked by remote side */
          ucs_queue_head_t       window;       /* send window */
-         uct_ud_ep_pending_op_t pending;      /* pending control ops */
+         uct_ud_ep_pending_op_t pending;      /* pending ops */
+         ucs_time_t             send_time;    /* tx time of last packet */
+         uct_ud_psn_t           rt_psn;       /* last psn that was retransmitted */
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: better but this psn together with other psn's to conserve memory used by field padding

File: src/uct/ib/ud/base/ud_ep.h [hunk 27]
```diff
@@ -98,6 +166,13 @@ struct uct_ud_ep {
     } rx;
     ucs_list_link_t          cep_list;
     uint32_t                 conn_id;      /* connection id. assigned in connect_to_iface() */
+    struct {
+        ucs_callback_t       *cb;
+        ucs_list_link_t      comp_list;
+    } comp;
+    ucs_wtimer_t slow_timer;
+    UCT_UD_EP_HOOK_DECLARE(timer_hook);
+    ucs_wtimer_t fast_timer;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: are we going to use fast timer?

File: src/uct/ib/ud/base/ud_ep.h [hunk 28]
```diff
@@ -98,6 +166,13 @@ struct uct_ud_ep {
     } rx;
     ucs_list_link_t          cep_list;
     uint32_t                 conn_id;      /* connection id. assigned in connect_to_iface() */
+    struct {
+        ucs_callback_t       *cb;
+        ucs_list_link_t      comp_list;
+    } comp;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: where is this used?

File: src/uct/ib/ud/base/ud_iface.c [hunk 29]
```diff
@@ -413,8 +440,8 @@ void uct_ud_iface_query(uct_ud_iface_t *iface, uct_iface_attr_t *iface_attr)
     iface_attr->cap.am.max_zcopy      = 0;
 
     iface_attr->cap.put.max_short     = iface->config.max_inline - sizeof(uct_ud_neth_t) - sizeof(uct_ud_put_hdr_t);
-    iface_attr->cap.am.max_bcopy      = mtu - sizeof(uct_ud_neth_t);
-    iface_attr->cap.am.max_zcopy      = 0;
+    iface_attr->cap.put.max_bcopy      = 0;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: indentation

File: src/uct/ib/ud/base/ud_iface.c [hunk 30]
```diff
@@ -475,64 +504,49 @@ void uct_ud_iface_replace_ep(uct_ud_iface_t *iface,
...
+void uct_ud_iface_reserve_skbs(uct_ud_iface_t *iface, int count)
+{
+    int i;
+    uct_ud_send_skb_t *skb;
 
+    for (i = 0; i < count; i++) {
+        skb = ucs_mpool_get(&iface->tx.mp);
+        if (skb == NULL) {
+            ucs_fatal("failed to reserve %d tx skbs", count);
+            return;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: no need  for return here

File: src/uct/ib/ud/base/ud_iface.h [hunk 31]
```diff
@@ -58,7 +69,9 @@ struct uct_ud_iface {
         ucs_mpool_t            mp;
         int16_t                available;
         unsigned               unsignaled;
-        ucs_queue_head_t       pending_ops;
+        /* pull of skbs that are reserved for retransmissions/ctl packects */
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.h]: pool

File: src/uct/ib/ud/base/ud_inl.h [hunk 32]
```diff
@@ -47,17 +36,33 @@ uct_ud_send_skb_t *uct_ud_iface_get_tx_skb(uct_ud_iface_t *iface,
     return skb;
 }
 
-/* same as function above but also check that iface is connected */
+/* same as above but also check ep resources: window&connection state */
 static UCS_F_ALWAYS_INLINE uct_ud_send_skb_t *
 uct_ud_ep_get_tx_skb(uct_ud_iface_t *iface, uct_ud_ep_t *ep)
 {
-    if (!uct_ud_ep_is_connected(ep)) {
+    if (ucs_unlikely(!uct_ud_ep_is_connected(ep) ||
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_inl.h]: can we unite these checks? this is data path.
can be in another commit too.

File: src/uct/ib/ud/base/ud_log.c [hunk 33]
```diff
@@ -105,6 +105,21 @@ static int uct_ud_dump_ep(char *p, int max, uct_ud_ep_t *ep)
     return n;
 } 
 
+static char *pkt_type2str(int type)
+{
+    switch(type) {
+        case UCT_AM_TRACE_TYPE_SEND:
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_log.c]: pls indent 'case' on same column as 'switch'

File: src/uct/ib/ud/base/ud_log.h [hunk 34]
```diff
@@ -30,5 +30,9 @@ void uct_ud_log_packet(const char *file, int line, const char *function,
                           UCT_AM_TRACE_TYPE_RECV, _neth, _len); \
     }
 
-
+#define uct_ud_iface_log_rxdrop(_iface, _ep, _neth, _len) \
+    if (ucs_log_enabled(UCS_LOG_LEVEL_TRACE_DATA)) { \
+        uct_ud_log_packet(__FILE__, __LINE__, __FUNCTION__, _iface, _ep, \
+                          UCT_AM_TRACE_TYPE_RECV_DROP, _neth, _len); \
+    }
 #endif
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_log.h]: need space line

File: test/gtest/Makefile.am [hunk 35]
```diff
@@ -95,7 +95,10 @@ gtest_SOURCES += \
     uct/test_ib.cc
 if HAVE_TL_UD
 gtest_SOURCES += \
+    uct/ud_base.cc \
```
Comment (yosefe) [inline test/gtest/Makefile.am]: indentation

File: test/gtest/uct/test_ud_slow_timer.cc [hunk 36]
```diff
@@ -0,0 +1,165 @@
...
+    ep(m_e2)->rx.rx_hook = uct_ud_ep_null_hook;
+    EXPECT_EQ(N+1, ep(m_e1)->tx.psn);
+    EXPECT_EQ(rx_limit, ucs_frag_list_sn(&ep(m_e2)->rx.ooo_pkts));
+    orig_avail = iface(m_e1)->tx.available;
+    /* allow only 6 outgoing packets. It will allow to get ack
+     * from receiver
+     */
+    iface(m_e1)->tx.available = 6;
+    twait(500);
+  //  short_progress_loop();
```
Comment (yosefe) [inline test/gtest/uct/test_ud_slow_timer.cc]: pls remove

File: src/uct/ib/ud/base/ud_inl.h (hunk 32 above)
Comment (alex-mikheev) [inline src/uct/ib/ud/base/ud_inl.h]: yes this is on todo list.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1071/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/499/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/501/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
looks good, pls squash

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1072/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/502/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/503/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1073/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/505/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1075/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/507/ for details (Mellanox internal link).
